#Minor adjustments from https://devonderaad.github.io/SNPfiltR/
#DeRaad, D.A. (2022), SNPfiltR: an R package for interactive and reproducible SNP filtering. Molecular Ecology Resources, 22, 2443–2453, 1–15. https://doi.org/10.1111/1755-0998.13618.
#Knaus, Brian J., and Niklaus J. Grunwald. 2017. VCFR: a package to manipulate and visualize variant call format data in R. Molecular Ecology Resources 17(1):44-53. https://doi.org/10.1111/1755-0998.12549.

library(vcfR)
library(SNPfiltR)

# load vcf file
vcf <- read.vcfR("populations.snps.vcf")

popmap <- read.table("ac_scandinavia_map.txt",header = T, sep = "\t", 
                     stringsAsFactors = F)

#visualize distributions
hard_filter(vcfR=vcf)

#depth and gq
vcf <- hard_filter(vcfR=vcf, depth = 5, gq = 30)

# 0.16% of genotypes fall below a read depth of 5 and were converted to NA
# 0% of genotypes fall below a genotype quality of 30 and were converted to NA

# allele balance
vcf <- filter_allele_balance(vcf)
# 12.79% of het genotypes (2.98% of all genotypes) fall outside of 0.25 - 0.75
# allele balance ratio and were converted to NA

# max depth

max_depth(vcf)
# cutoff is not specified, exploratory visualization will be generated.
# dashed line indicates a mean depth across all SNPs of 64.6

vcf <- max_depth(vcf,maxdepth = 130)
# maxdepth cutoff is specified, filtered vcfR object will be returned
# 1.99% of SNPs were above a mean depth of 130 and were removed from the vcf

# missing data
missing_by_sample(vcfR=vcf, popmap = popmap)

vcf <- missing_by_sample(vcfR=vcf, cutoff = .8)
#13 samples are above a 0.8 missing data cutoff, and were removed from VCF

#subset popmap to only include retained individuals
popmap <- popmap[popmap$id %in% colnames(vcf@gt),]

#remove invariant sites generated by dropping individuals
vcf <- min_mac(vcf, min.mac = 1)
#5.57% of SNPs fell below a minor allele count of 1 and were removed from the VCF

#verify that missing data is not driving clustering patterns among the retained samples
miss_sample <- assess_missing_data_pca(vcfR=vcf, popmap = popmap, thresholds = .8, clustering = FALSE)

prob_sample <- as.data.frame(miss[[1]][7])

max(prob_sample["missing"])
which.max(prob_sample[["missing"]])

#removed ac_norway_014
vcf <- vcf[,colnames(vcf@gt) != "ac_norway_014"]
popmap <- popmap[popmap$id %in% colnames(vcf@gt),]

# missing by SNP
missing_by_snp(vcf)

miss_snp <- assess_missing_data_pca(vcfR=vcf, popmap = popmap, thresholds = c(.7,.8,.85), clustering = FALSE)

#apply cutoff 0.8
vcf <- missing_by_snp(vcf, cutoff = .80)

# MAC 
vcf.mac<-min_mac(vcfR = vcf, min.mac = 2)
rm(vcf.mac)

#t-SNE
miss_tsne <- assess_missing_data_tsne(vcf, popmap, clustering = T)

#plot depth per snp and per sample
dp <- extract.gt(vcf, element = "DP", as.numeric=TRUE)
heatmap.bp(dp, rlabels = FALSE)

#plot genotype quality per snp and per sample
gq <- extract.gt(vcf, element = "GQ", as.numeric=TRUE)
heatmap.bp(gq, rlabels = FALSE)

# save filtered vcf
vcfR::write.vcf(vcf, "ac_scandinavia_filt.vcf")

#linkage filter vcf to thin SNPs to one per 500bp
vcf.thin <- distance_thin(vcf, min.distance = 500)

#write out thinned vcf
vcfR::write.vcf(vcf.thin, "ac_scandinavia_filt_thin.vcf.gz")







